lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shared.hs

```diff
@@ -0,0 +1,178 @@
+    ( Hash (..) )
+import Control.DeepSeq
+    ( NFData (..) )
+import Control.Monad
+    ( (<=<) )
+import Crypto.Hash
+    ( hash )
+import Data.ByteString
+    ( ByteString )
+import GHC.Generics
+    ( Generic )
+
+{-------------------------------------------------------------------------------
+                            Sequential Derivation
+-------------------------------------------------------------------------------}
+
+-- | A cryptographic key for Shared address derivation, with phantom-types to
+-- disambiguate derivation paths
+--
+-- @
+-- let rootPrivateKey = SharedKey 'RootK XPrv
+-- let accountPubKey = SharedKey 'AccountK XPub
+-- let addressPubKey = SharedKey 'AddressK XPub
+-- @
+newtype SharedKey (depth :: Depth) key =
+    SharedKey { getKey :: key }
+    deriving stock (Generic, Show, Eq)
+
+instance (NFData key) => NFData (SharedKey depth key)
+
+-- | Generate a root key from a corresponding seed.
+-- The seed should be at least 16 bytes.
+generateKeyFromSeed
+    :: (SomeMnemonic, Maybe SomeMnemonic)
+       -- ^ The actual seed and its recovery / generation passphrase
+    -> Passphrase "encryption"
+    -> SharedKey 'RootK XPrv
+generateKeyFromSeed = unsafeGenerateKeyFromSeed
+
+-- | Generate a new key from seed. Note that the @depth@ is left open so that
+-- the caller gets to decide what type of key this is. This is mostly for
+-- testing, in practice, seeds are used to represent root keys, and one should
+-- use 'generateKeyFromSeed'.
+unsafeGenerateKeyFromSeed
+    :: (SomeMnemonic, Maybe SomeMnemonic)
+        -- ^ The actual seed and its recovery / generation passphrase
+    -> Passphrase "encryption"
+    -> SharedKey depth XPrv
+unsafeGenerateKeyFromSeed mnemonics pwd =
+    SharedKey $ unsafeGenerateKeyFromSeedShelley mnemonics pwd
+
+instance HardDerivation SharedKey where
+    type AddressIndexDerivationType SharedKey = 'Soft
+
+    deriveAccountPrivateKey pwd (SharedKey rootXPrv) ix =
+        SharedKey $ deriveAccountPrivateKeyShelley purposeCIP1854 pwd rootXPrv ix
+
+    deriveAddressPrivateKey pwd (SharedKey accXPrv) role ix =
+        SharedKey $ deriveAddressPrivateKeyShelley pwd accXPrv role ix
+
+instance SoftDerivation SharedKey where
+    deriveAddressPublicKey (SharedKey accXPub) role ix =
+        SharedKey $ deriveAddressPublicKeyShelley accXPub role ix
+
+{-------------------------------------------------------------------------------
+                            WalletKey implementation
+-------------------------------------------------------------------------------}
+
+instance WalletKey SharedKey where
+    changePassphrase (Passphrase oldPwd) (Passphrase newPwd) (SharedKey prv) =
+        SharedKey $ xPrvChangePass oldPwd newPwd prv
+
+    publicKey (SharedKey prv) =
+        SharedKey (toXPub prv)
+
+    digest (SharedKey pub) =
+        hash (unXPub pub)
+
+    getRawKey =
+        getKey
+
+    liftRawKey =
+        SharedKey
+
+    keyTypeDescriptor _ =
+        "shared"
```

Comment:
I have doubts what should be here to be honest

Reply:
This is the name then used to prefix database's names. In other wallet scheme, we typically use a 3-letter identifier:

- she
- ica
- rnd

I'd say that for consistency, we probably want to also have a 3-letter name here but I am perhaps being too maniac :grimacing:

Reply:
went for `sha`